client/js/lounge.js

```diff
@@ -63,128 +62,8 @@ $(function() {
-			offset.left -= menuWidth - that.outerWidth();
-			offset.top += that.outerHeight();
-			return offset;
-		}
-
-		offset = {left: e.pageX, top: e.pageY};
-
-		if ((window.innerWidth - offset.left) < menuWidth) {
-			offset.left = window.innerWidth - menuWidth;
-		}
-
-		if ((window.innerHeight - offset.top) < menuHeight) {
-			offset.top = window.innerHeight - menuHeight;
-		}
-
-		return offset;
-	}
-
-	function showContextMenu(that, e) {
-		const target = $(e.currentTarget);
-		let output = "";
-
-		if (target.hasClass("user")) {
-			output = templates.contextmenu_item({
-				class: "user",
-				action: "whois",
-				text: target.text(),
-				data: target.data("name"),
-			});
-			output += templates.contextmenu_divider();
-			output += templates.contextmenu_item({
-				class: "action-whois",
-				action: "whois",
-				text: "User information",
-				data: target.data("name"),
-			});
-			output += templates.contextmenu_item({
-				class: "action-query",
-				action: "query",
-				text: "Direct messages",
-				data: target.data("name"),
-			});
-
-			const channel = target.closest(".chan");
-
-			if (utils.hasRoleInChannel(channel, ["op"]) && channel.data("type") === "channel") {
-				output += templates.contextmenu_divider();
-				output += templates.contextmenu_item({
-					class: "action-kick",
-					action: "kick",
-					text: "Kick",
-					data: target.data("name"),
-				});
-			}
-		} else if (target.hasClass("chan")) {
-			let itemClass;
-
-			if (target.hasClass("lobby")) {
-				itemClass = "network";
-			} else if (target.hasClass("query")) {
-				itemClass = "query";
-			} else {
-				itemClass = "chan";
-			}
-
-			output = templates.contextmenu_item({
-				class: itemClass,
-				action: "focusChan",
-				text: target.attr("aria-label"),
-				data: target.data("target"),
-			});
-			output += templates.contextmenu_divider();
-
-			if (target.hasClass("lobby")) {
-				output += templates.contextmenu_item({
-					class: "list",
-					action: "list",
-					text: "List all channels",
-					data: target.data("id"),
-				});
-				output += templates.contextmenu_item({
-					class: "join",
-					action: "join",
-					text: "Join a channel…",
-					data: target.data("id"),
-				});
-			}
-
-			if (target.hasClass("channel")) {
-				output += templates.contextmenu_item({
-					class: "list",
-					action: "banlist",
-					text: "List banned users",
-					data: target.data("id"),
-				});
-			}
-
-			output += templates.contextmenu_item({
-				class: "close",
-				action: "close",
-				text: target.hasClass("lobby") ? "Disconnect" : target.hasClass("channel") ? "Leave" : "Close",
-				data: target.data("target"),
-			});
-		}
-
-		contextMenuContainer.show();
-		contextMenu
-			.html(output)
-			.css(positionContextMenu($(that), e));
-
-		return false;
-	}
-
 	viewport.on("contextmenu", ".network .chan", function(e) {
-		return showContextMenu(this, e);
+		return contextMenuFactory.createContextMenu(this, e).show();
```

Comment:
Correct me if I'm wrong: I feel like you'd be better of passing `$(this)` around (the repetition of `$(this)` vs. `this` doesn't warrant abstracting that IMO, I'd rather consistently pass jQuery objects around in your `contextMenuFactory` method and `new ContextMenu`), then in [here](https://github.com/thelounge/thelounge/pull/1878/files#diff-d15cea7cb8a0229dd9b492ffa0d7ac15R59) you can do:

```js
.css(positionContextMenu(selectedElement, event));
```

(i.e. use the jQuery object itself rather than wrapping it with `$()` every time)

Reply:
Yeah, sure